### PR TITLE
NIFI-8732: Adding a HashiCorp Vault K/V Stateless Param Provider

### PR DESCRIPTION
**Pull request by gresockj**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR
**HashiCorpVaultParameterValueProvider**

This provider reads parameter values from HashiCorp Vault, and expects secrets to exist in
the Key/Value (unversioned) Secrets Engine.  The connection to a Vault server can be configured
via the `./conf/bootstrap-hashicorp-vault.conf` file, which comes with NiFi.

An example of creating a single secret in the correct format is:

```
vault kv put "nifi-kv/Context/param" value=my-vault-value
```

In this example, `nifi-kv` would be supplied by the `vault.kv.path` property in the `bootstrap-hashicorp-vault.conf` file, 
`Context` is the name of a Parameter Context, and `param` is the name of the parameter whose value should be retrieved from the Vault server.

This Parameter Provider requires the following properties:

| Property Name | Description | Example Value |
|---------------|-------------|---------------|
| nifi.stateless.parameter.provider.\<key>.properties.vault-configuration-file | The filename of a configuration file specifying the Vault settings | ./conf/bootstrap-hashicorp-vault.conf |

An example of configuring this provider in the dataflow configuration file is:

```
nifi.stateless.parameter.provider.Vault.name=HashiCorp Vault Provider
nifi.stateless.parameter.provider.Vault.type=org.apache.nifi.stateless.parameter.HashiCorpVaultParameterValueProvider
nifi.stateless.parameter.provider.Vault.properties.vault-configuration-file=./conf/bootstrap-hashicorp-vault.conf
```

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-hashicorp-vault-bundle/nifi-hashicorp-vault-parameter-value-provider/pom.xml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-vault-utils</artifactId>
+            <version>1.15.0-SNAPSHOT</version>
+        </dependency>
+        <!-- test dependencies -->
+        <dependency>
+            <groupId>commons-logging</groupId>
+            <artifactId>commons-logging</artifactId>
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-hashicorp-vault-bundle/nifi-hashicorp-vault-parameter-value-provider/pom.xml:39]: The `jcl-over-slf4j` dependency should be used in place of `commons-logging` to maintain SLF4J as the standard logger throughout the system.

File: nifi-assembly/pom.xml [hunk 2]
```diff
@@ -760,6 +760,12 @@ language governing permissions and limitations under the License. -->
             <version>1.15.0-SNAPSHOT</version>
             <type>nar</type>
         </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-hashicorp-vault-nar</artifactId>
```
Comment (exceptionfactory) [inline nifi-assembly/pom.xml:765]: What do you think of naming this to something more specific? It seems like there could be other libraries for interacting with HashiCorp Vault, so including `parameter-value-provider` in the name would be helpful.

Comment (gresockj) [reply nifi-assembly/pom.xml:765]: My thought was that this nar would also contain jars for a `hashicorp-vault-parameter-provider` and a `hashicorp-vault-controller-service`.  In light of that, I was considering this more analogous to the `nifi-aws-nar`.  What do you think?

Comment (exceptionfactory) [reply nifi-assembly/pom.xml:765]: Thanks for the reply, that seems reasonable, and it could always separated later if needed.

**Review by exceptionfactory — APPROVED** [review]
Thanks for the contribution @gresockj, verified parameter resolution in a stateless flow using HashiCorp Vault server, nice work! +1 Merging.

**Pull request closed without merging**